src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,152 @@
+  }
+
+  // Spins up the shooter, and requests feeding it when the rollers are within parameters.
+  // Receives distance-to-target from Limelight, or other sensor.
+  // DOES NOT WORK CURRENTLY!
+  public ParallelCommandGroup shooterFire() {
+    addCommands(
+      setShooterSpeeds(calculateShooterSpeeds(LimelightGetDistance())),
+      isShooterReady(LimelightGetDistance())).withName("Shooter fire command");
+  }
+
+  // Idle shooter command; for default command purposes
+  public Command shooterIdle() {
+    return run(
+        () -> {
+          setShooterSpeeds(Constants.ShooterConstants.SHOOTER_IDLE_SPEEDS);
+        }).withName("Idle shooter command");
+  }
+
+  // Checks if shooter is ready. 
+  public void isShooterReady(double distance) {
+    double targetSpeedTop = calculateShooterSpeeds(distance)[0];
+    double targetSpeedBottom = calculateShooterSpeeds(distance)[1];
+
+    double minSpeedToleranceTop = targetSpeedTop * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceTop = targetSpeedTop * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double minSpeedToleranceBottom = targetSpeedBottom * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceBottom = targetSpeedBottom * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+
+    double speedTop = shooterTopMotor.getEncoder().getVelocity();
+    double speedBottom = shooterBottomMotor.getEncoder().getVelocity();
+
+    if ((minSpeedToleranceTop < speedTop && speedTop < maxSpeedToleranceTop)
+      && (minSpeedToleranceBottom < speedBottom && speedBottom < maxSpeedToleranceBottom) == true) {
+        // Request note from feeder.
+    }
+  }
+
+  // Speed in RPM. Top is index 0, bottom is index 1.
+  public void setShooterSpeeds(double[] speeds) {
+    shooterTopMotor.getPIDController().setReference(speeds[0], CANSparkBase.ControlType.kVelocity);
+    shooterBottomMotor
+        .getPIDController()
+        .setReference(speeds[1], CANSparkBase.ControlType.kVelocity);
+  }
+
+  // Top is index 0, bottom is index 1.
+  public static double[] calculateShooterSpeeds(double distance) {
+    double[] speeds = {0, 0};
+    double distanceProportion;
+
+    // If distance below minimum, set speed to minimum.
+    if (distance < speakerConfigList[0].getDistance()) {
+      speeds[0] = speakerConfigList[0].getSpeedTop();
+      speeds[1] = speakerConfigList[0].getSpeedBottom();
+    } else {
+      // A linear search which determines which points the input distance falls between. May be
+      // converted to a binary search if there are many points
+      for (int i = 0; i < speakerConfigList.length; i++) {
+        // If distance above maximum, set speed to maximum.
+        if (i == speakerConfigList.length - 1) {
+          speeds[0] = speakerConfigList[i].getSpeedTop();
+          speeds[1] = speakerConfigList[i].getSpeedBottom();
+          break;
```

Comment:
IMO these `break` statements are unnecessary. Instead consider just `return`ing at these points in the flow.

If you do that, you could then have the loop go up to length-1 the pull this `if` block out of the loop, which I think would be more readable.

Reply:
Doing so does not simplify the logic or, really, make it more readable; you're just shifting the code around. It just ends up adding more, similarly arcane lines in the middle of the method.

Reply:
"Resolved" unless this change is viewed as critical